drv/sprot-api/src/lib.rs

```diff
@@ -3,241 +3,243 @@
+            .unwrap_lite();
+
+        // Copy the blob into the buffer after the serialized body
+        blob.read_range(0..blob.len(), &mut buf[Header::MAX_SIZE + size..])
+            .map_err(|_| SprotProtocolError::ServerRestarted)?;
+
+        size += blob.len();
+
+        // Create a header, now that we know the size of the body
+        let header = Header::new(size.try_into().unwrap_lite());
+
+        // Serialize the header
+        size += hubpack::serialize(buf, &header).unwrap_lite();
+
+        // Compute and serialize the CRC
+        let crc = CRC16.checksum(&buf[..size]);
+        size += hubpack::serialize(&mut buf[size..], &crc).unwrap_lite();
+
+        Ok(size)
+    }
+
+    // Deserialize and return a `Msg`
+    pub fn unpack(buf: &'a [u8]) -> Result<Msg<'a, T, N>, SprotProtocolError> {
+        let (header, rest) = hubpack::deserialize::<Header>(buf)?;
+        if header.protocol != Protocol::V2 {
+            return Err(SprotProtocolError::UnsupportedProtocol);
+        }
+        Self::unpack_body(header, buf, rest)
+    }
+
+    /// Deserialize just the body, given a header that was already deserialized.
+    pub fn unpack_body(
+        header: Header,
+        // The buffer containing the entire serialized `Msg` including the `Header`
+        buf: &[u8],
+        // The part of the after the header buffer including the body and CRC
+        rest: &'a [u8],
+    ) -> Result<Msg<'a, T, N>, SprotProtocolError> {
+        let (body, blob_buf) = hubpack::deserialize::<T>(rest)?;
+        let end = Header::MAX_SIZE + header.body_size as usize;
+        let (checksummed_part, tail) = buf.split_at(end);
+        let computed_crc = CRC16.checksum(checksummed_part);
+
+        // The CRC comes after the body, and is not included in header body_len
+        let (crc, _) = hubpack::deserialize(tail)?;
+
+        if computed_crc == crc {
+            let blob_len =
+                header.body_size as usize - (rest.len() - blob_buf.len());
+            let blob = &blob_buf[..blob_len];
+            Ok(Msg { header, body, blob })
+        } else {
+            Err(SprotProtocolError::InvalidCrc)
         }
     }
 }
 
-impl From<hubpack::Error> for SprotError {
-    fn from(_: hubpack::Error) -> Self {
-        SprotError::Serialization
-    }
+/// Protocol version
+/// This is the first byte of any Sprot request or response
+#[derive(
+    Copy, Clone, Eq, PartialEq, Deserialize, Serialize, SerializedSize,
+)]
+#[repr(u8)]
+pub enum Protocol {
+    /// Indicates that no message is present.
+    Ignore,
+    /// The first sprot format with hand-rolled serialization.
+    V1,
+    /// The second format, using hubpack
+    V2,
 }
 
-impl From<HandoffDataLoadError> for SprotError {
-    fn from(_: HandoffDataLoadError) -> Self {
-        SprotError::Stage0HandoffError
-    }
+/// The body of a sprot request
+#[derive(Clone, Serialize, Deserialize, SerializedSize)]
+pub enum ReqBody {
+    Status,
```

Comment:
We need a Status message that allows us to change protocol later.
What is the necessary information?
  - protocols supported.
  - buffer sizes?
  - Maybe ROM crc (not because this is the right place, but because NXP doesn't give us an explicit HW version and do we have a different API for that?
  talk more later.

  -

Reply:
I made a few related changes:
 1. Status now contains just the current and min versions for each of the RoT and SP along with buffer sizes
 2. I added a long doc comment explaining how extending the protocol should work